From f63fb4d77dcbd324d5c5ca749bf454bb981aa6d9 Mon Sep 17 00:00:00 2001
From: Niko Tyni <ntyni@debian.org>
Date: Mon, 4 Aug 2014 10:32:36 +0300
Subject: [PATCH 1/3] Define MP_IN_XS to avoid referencing &perl_module outside
 of mod_perl

APLOG_USE_MODULE(perl) expands to

 extern module perl_module; static int * const aplog_module_index = &(perl_module.module_index);

which makes loading APR.so impossible outside mod_perl unless
aplog_module_index is optimized away by the compiler.
This causes mod_perl test suite failures in t/apr-ext at "gcc -O0".

See also
 http://svn.apache.org/viewvc?view=revision&revision=68792
 http://svn.apache.org/viewvc?view=revision&revision=1410295

Upstream thread at
 http://mail-archives.apache.org/mod_mbox/perl-dev/201408.mbox/%3C20140807115045.GA6427%40estella.local.invalid%3E

Bug-Debian: https://bugs.debian.org/756989
---
 xs/APR/APR/Makefile.PL    | 7 +++++++
 xs/APR/PerlIO/Makefile.PL | 3 +++
 2 files changed, 10 insertions(+)

diff --git a/xs/APR/APR/Makefile.PL b/xs/APR/APR/Makefile.PL
index 8745a54..d569dde 100644
--- a/xs/APR/APR/Makefile.PL
+++ b/xs/APR/APR/Makefile.PL
@@ -25,6 +25,13 @@ $libs = delete $args{LIBS} if $args{LIBS};
 
 my $build = ModPerl::BuildMM::build_config();
 
+my $ccopts = $build->ccopts;
+
+# avoid referencing &perl_module outside of mod_perl
+$ccopts .= ' -DMP_IN_XS';
+
+$args{CCFLAGS} = $ccopts;
+
 my @apru_link_flags = $build->apru_link_flags;
 $libs .= join ' ', @apru_link_flags if @apru_link_flags;
 
diff --git a/xs/APR/PerlIO/Makefile.PL b/xs/APR/PerlIO/Makefile.PL
index 4a8f60d..ca102bb 100644
--- a/xs/APR/PerlIO/Makefile.PL
+++ b/xs/APR/PerlIO/Makefile.PL
@@ -25,6 +25,9 @@ if ($build->has_large_files_conflict) {
         : '';
 }
 
+# avoid referencing &perl_module outside of mod_perl
+$ccopts .= ' -DMP_IN_XS';
+
 ModPerl::BuildMM::WriteMakefile(
     NAME         => 'APR::PerlIO',
     VERSION_FROM => 'PerlIO.pm',
-- 
2.1.0.rc1

